lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->configuration->get(PreloadHeroImageConfiguration::PRELOAD_HERO_IMAGE) === false) {
+            return;
+        }
+
+        $heroImages    = $this->findHeroImages($document);
+        $referenceNode = $document->viewport;
+
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > self::DATA_HERO_MAX) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = self::DATA_HERO_MAX;
+        }
+
+        $isAmpStory = Amp::isAmpStory($document);
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->generatePreload($heroImages[$index], $document, $errors, $referenceNode);
+            if (! $isAmpStory) {
+                // AMP Stories don't support SSR'd <amp-img> yet.
+                // See https://github.com/ampproject/amphtml/issues/29850.
+                $this->generateImg($heroImages[$index], $document);
+            }
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document Document to look for hero images in.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document)
+    {
+        $heroImageCandidate = null;
+        $heroImages         = [];
+        $node               = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
+            }
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = $this->skipNodeAndChildren($node);
+            } else {
+                $node = $this->nextNode($node);
+            }
+        }
+
+        // Optimize data-hero element if defined.
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Fall back to auto-detected hero image if available.
+        if ($heroImageCandidate) {
+            return [$heroImageCandidate];
+        }
+
+        // No hero images to optimize.
+        return [];
+    }
+
+    /**
+     * Detect a hero image with the data-hero attribute.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image, or null if none detected.
+     */
+    private function detectImageWithDataHero(DOMElement $element)
+    {
+        if (
+            $element->tagName === Extension::IMAGE
```

Comment:
Question: Why `tagName` and not `nodeName`? I see both are being used in the Optimizer here. I can't see any difference in behavior. The sanitizers usually use `nodeName` though I see there are a few instances of `tagName` as well.

Reply:
`tagName` is "the name of the tag/element", but only exists for `DOMElement` objects.

I usually prefer to use `tagName` when I know I'm dealing with elements because it will throw an actual error when I happen to call it on a non-element: https://3v4l.org/RT5Zj. `nodeName` will always return "something", which could obfuscate a bug.

I don't really feel too strongly about this, though, if you think this is confusing.

Reply:
No, that's great reasoning. I'll start using `tagname` as well.